### DISCUSSION TITLE
Listbox :  Multi/Single select property

### DISCUSSION THREAD

**Discussion by bernardLescure** [joonaspaakko/ScriptUI-Dialog-Builder-Joonas#94] [Q&A]
Hi, 
in the super power UI editor (👍 ), 
we've got a checkbox option for / Lisbox => creation properties =>  multiselect.

This option seems not to have effect on the CEP export, 
In the header script object I can see  
`<!--  .... "creationProps":{"multiselect":false, ... -->`
But no difference in CEP / html export if **multiselect** is **on** or **off**.
Is there a way to change this property ? 

Rather than adding a list of buttons, I would prefer using a list box with single selection.
I'm just starting to use it for AnimateCC panels. 
Thank you again for this super tool !  🥇 

Nardo

**Comment by joonaspaakko** [comment]
Right now there is no ideal way to do it, but there is a way to do it. Basically you just have to self manage it so that only one item is selected at all times... which is unfortunately not as easy as it could be.

Here's a [codepen](https://codepen.io/joonaspaakko/pen/MWvaYLR?editors=0010).

It should be pretty self sufficient in that I made it listen to all `listbox` type items.

```js
var getElIndex = function( el ) {
  return [...el.parentNode.children].indexOf(el);
};

var listBoxes = SDB.typeAll('listbox');

listBoxes.forEach(function( listbox ) {
  listbox.addEventListener("onSelect", function(e) {
    
    var currentEl = e.detail.element;
    var selectionArray = e.detail.selection.index;
    var currentIndex;
    
    if ( selectionArray.length > 1 ) {
      currentIndex = getElIndex(currentEl.parentNode);
      selectionArray.forEach(function( index ) {
        if ( currentIndex !== index ) SDB.Listbox.deselect( listbox, index);
      });
    }
    else {
      currentIndex = selectionArray[0];
    }
    
    listBoxSelected( listbox, currentEl, currentIndex );
    
  });
});

function listBoxSelected( lisbox, item, index ) {
  if ( index === 1 ) { 
    alert('Index: ' + index) 
  }
}
```

-----

Alternatively, and perhaps the better option in your case would be to remove all events from list boxes and then it would be way easier to impose your own click events whichever way you want.

Here's a [codepen](https://codepen.io/joonaspaakko/pen/ZEJbGVO?editors=0010) link for that.

```js
var listBoxes = SDB.typeAll('listbox');
listBoxes.forEach(function( listbox ) {
  listbox.outerHTML = listbox.outerHTML;
});


var listBoxes = SDB.typeAll('listbox');
listBoxes.forEach(function( listbox ) {
  listbox.querySelectorAll('ul li').forEach(function( item ) {
    item.addEventListener('click', function( e ) {
      
      console.log( e.target.textContent );
      
    });
  });
});
```

**Reply by joonaspaakko** [reply]
And to be clear why I posted these two code snippets. 

- **Code n.1:** Achieves single select with no compromises other than the code being a little unwieldy and you likely won't have performance issues, but it's pretty inefficient because it has to be done it in a very roundabout way.
 
- **Code n.2:** Removes the built-in listbox interactivity entirely, so it doesn't have to do so much on every click. The downside is that it doesn't actually select items on click, but if the goal is to have it work only as clickable buttons, then that shouldn't be a problem. I attached a new click event on every item as an example, which you could use to make the buttons do stuff. And you could, if needed, add and/remove `selected` class in `<li>` items... but at that point it might be better to just use code n.1.

**Reply by bernardLescure** [reply]
!!!! Yess

I had looked into the first solution and I think it will do the job, 

In other context, I've been digging around this : aria-multiselectable (property) of listbox
[https://www.w3.org/TR/2017/WD-wai-aria-practices-1.1-20170628/examples/listbox/listbox.html](https://www.w3.org/TR/2017/WD-wai-aria-practices-1.1-20170628/examples/listbox/listbox.html)
But don't understand anything about aria (for me it's my guitar Brand ... :-(  ) ... 

Thank you again 

I 'll keep you posted

**Reply by joonaspaakko** [reply]
I feel like the name I've heard the most for these is `dual list box` and possibly in some cases something like `transfer list` or `transfer box`... Or just `transfer component`, but the first one is probably the most commonly used. It isn't too difficult to do yourself with SDB's SUI elements, but depending on your needs, it might be easier to just add an existing script... or possibly from a UI component library. 

I noticed I've apparently forgotten to add a remove() method for listboxes (it's in the docs but the method doesn't exist), so I made a small example where I extended `SDB.Listbox` with `SDB.Listbox.remove( element, textString )` and added basic dual list box functionality. Here's the [codepen link](https://codepen.io/joonaspaakko/pen/gOxPMRy?editors=0010). It might require way more work to be fully functional for your purposes but it's an ok start at least.

```js
var getElIndex = function( el ) {
  return [...el.parentNode.children].indexOf(el);
};

var listBoxes = SDB.typeAll('listbox');

listBoxes.forEach(function( listbox ) {
  listbox.addEventListener("onSelect", function(e) {
    
    var currentEl = e.detail.element;
    var selectionArray = e.detail.selection.index;
    var currentIndex;
    
    if ( selectionArray.length > 1 ) {
      currentIndex = getElIndex(currentEl.parentNode);
      selectionArray.forEach(function( index ) {
        if ( currentIndex !== index ) SDB.Listbox.deselect( listbox, index);
      });
    }
    else {
      currentIndex = selectionArray[0];
    }
    
    // listBoxSelected( listbox, currentEl, currentIndex );
    
  });
});

// Adding a missing method for item removal using text content
SDB.Listbox.remove = function( element, text ) {

  var isArray = Array.isArray( text );
  if ( !isArray ) text = [text];
  var listboxItems = pickAll('ul > li', element);
  listboxItems.forEach(function( item ) {
    text.forEach(function( textContent ) {
      if ( item.textContent === textContent ) item.remove();
    });
  });
  
};

// DUAL LIST BOX

var dualListbox = {
  left: SDB.name('listboxLeft'),
  right: SDB.name('listboxRight'),
  button: {
    left: SDB.name('addLeft'),
    right: SDB.name('addRight'),
  },
};


dualListbox.click = function ( side ) {
  dualListbox.button[ side ].addEventListener('click', function() {
    
    var inputSide = dualListbox[ side === 'left' ? 'right' : 'left' ];
    var inputText = SDB.Listbox.selection.text( inputSide )[0];
    var inputIndex = SDB.Listbox.selection.index( inputSide )[0];
    var outputSide = dualListbox[ side ];
    
    SDB.Listbox.remove( inputSide, inputText ); // Remove from input
    SDB.Listbox.add( outputSide, inputText ); // Add to output
      
    // Since the selected input item was removed, select another item...
    var listboxItems = pickAll('ul > li', inputSide);
    var lastIndex = listboxItems.length-1;
    var newInputIndex = (inputIndex <= lastIndex) ? inputIndex : lastIndex;
    SDB.Listbox.select( inputSide, newInputIndex );

  });
      
};

// Attaches the click event to each chevron button
dualListbox.click('left');
dualListbox.click('right');

```
